packages/scorer/src/provider/model-graded/faithfulness.ts

```diff
@@ -0,0 +1,156 @@
+3. John is a dedicated student.
+Explanation: The prompt states that he spends a significant amount of time studying and completing assignments. Additionally, it mentions that he often stays late in the library to work on his projects, which implies dedication. Verdict: Yes.
+4. John has a part-time job.
+Explanation: There is no information given in the context about John having a part-time job. Therefore, it cannot be deduced that John has a part-time job.  Verdict: No.
+5. John is interested in computer programming.
+Explanation: The context states that John is pursuing a degree in Computer Science, which implies an interest in computer programming. Verdict: Yes.
+Final verdict for each statement in order: No. No. Yes. No. Yes.
+
+Context:\n{{context}}
+{{statements}}
+Answer:
+`;
+
+export const checkLlmFaithfulness: ScoringFn = async ({ sample, config }) => {
+  let context = "";
+  if (config.type !== "llm-faithfulness") {
+    return [
+      {
+        name,
+        score: 0,
+        message: "invalid scoring function detected",
+      },
+    ];
+  }
+  if (config.context) {
+    let replacements: any = { ...sample.inputs };
+    if (sample.expected) {
+      replacements.expected = sample.expected;
+    }
+    context = replacePlaceholders(config.context, replacements);
+  }
+
+  // TODO: Figure out how to use placeholders for this scorer
+  // Output is not used currently, nor are claims replaced with placeholders
+  const messages: OpenAI.ChatCompletionMessageParam[] = [
+    {
+      role: "user",
+      content: replacePlaceholders(prompt, {
+        context,
+        statements: config.claims.reduce((agg, claim, index) => {
+          return `${agg}\nstatement ${index + 1}: ${claim}`;
+        }, ""),
+      }),
+    },
+  ];
+
+  try {
+    const response = await askLlmForEvalResult(messages, config.claims.length);
+    const totalScore = Object.keys(response).reduce(
+      (score, key) => score + (response[key] === "Yes" ? 1 : 0),
+      0,
+    );
+    const reasonsForNo = Object.keys(response).reduce(
+      (reasons: string[], key) =>
+        response[key] === "No"
+          ? [...reasons, response[`explanation_${key.split("_").pop()}`]!]
+          : reasons,
+      [],
+    );
+    return [
+      {
+        score: totalScore / config.claims.length,
+        name: config.name || name,
+        message: reasonsForNo.join("\n"),
+      },
+    ];
+  } catch (err) {
+    return [
+      {
+        score: 0,
+        name: config.name || name,
+        message:
+          (err as Error).message ||
+          "Failed to call LLM using @empiricalrun/ai package",
+      },
+    ];
+  }
+};
+
+async function askLlmForEvalResult(
+  messages: OpenAI.ChatCompletionMessageParam[],
+  numStatements: number,
+): Promise<{ [key: string]: string }> {
+  const ai = new EmpiricalAI("openai");
```

Comment:
**Suggestion:** The `askLlmForEvalResult` function is responsible for interacting with the LLM and processing the response. This function could be refactored to make it more testable and maintainable. For example, you could extract the logic for creating the tool function and the tool choice into separate functions, and make the function more modular by allowing the caller to pass in the required parameters. [Refactoring]
```suggestion
async function askLlmForEvalResult(
  messages: OpenAI.ChatCompletionMessageParam[],
  numStatements: number,
  ai: EmpiricalAI = new EmpiricalAI("openai"),
): Promise<{ [key: string]: string }> {
  const properties = createEvaluatorResponseProperties(numStatements);
  const completion = await ai.chat.completions.create({
    messages,
    model: "gpt-3.5-turbo",
    temperature: 0.1,
    tools: [createEvaluatorResponseTool(properties)],
    tool_choice: createEvaluatorResponseToolChoice(),
  });
  const rawResponse = completion.choices[0]!;
  const response = rawResponse.message.tool_calls![0];
  return JSON.parse(response!.function.arguments);
}

function createEvaluatorResponseProperties(numStatements: number): { [key: string]: any } {
  return [...Array(numStatements).keys()].reduce((agg: any, index) => {
    // ...
    return agg;
  }, {});
}

function createEvaluatorResponseTool(properties: { [key: string]: any }): {
  type: "function";
  function: {
    name: string;
    description: string;
    parameters: {
      type: "object";
      properties: { [key: string]: any };
      required: string[];
    };
  };
} {
  return {
    type: "function",
    function: {
      name: "set_evaluator_response",
      description: "Sets the response of the evaluation",
      parameters: {
        type: "object",
        properties,
        required: Object.keys(properties),
      },
    },
  };
}

function createEvaluatorResponseToolChoice(): {
  type: "function";
  function: { name: string };
} {
  return {
    type: "function",
    function: { name: "set_evaluator_response" },
  };
}
```